pkg/uio/buffer.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2018 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package uio
+
+import (
+	"encoding/binary"
+	"io"
+)
+
+// Marshaler is the interface implemented by an object that can marshal itself
+// into binary form.
+//
+// Marshal appends data to the buffer b.
+type Marshaler interface {
+	Marshal(b *Buffer)
+}
+
+// Unmarshaler is the interface implemented by an object that can unmarshal a
+// binary representation of itself.
+//
+// Unmarshal consumes data from the buffer b.
+type Unmarshaler interface {
+	Unmarshal(b *Buffer) error
+}
+
+// Buffer implements functions to manipulate byte slice buffers in specific
+// byte orders.
+type Buffer struct {
+	// data is the underlying data.
+	data []byte
+
+	// order is the byte order to write in / read in.
+	order binary.ByteOrder
+}
+
+// NewBEBuffer consumes b for marshaling or unmarshaling in big endian.
+func NewBEBuffer(b []byte) *Buffer {
+	return &Buffer{
+		data:  b,
+		order: binary.BigEndian,
+	}
+}
+
+// NewBuffer consumes b for marshaling or unmarshaling in the given byte order.
+func NewBuffer(b []byte, order binary.ByteOrder) *Buffer {
+	return &Buffer{
+		data:  b,
+		order: order,
+	}
+}
+
+// append appends n bytes to the Buffer and returns a slice pointing to the
+// newly appended bytes.
+func (b *Buffer) append(n int) []byte {
+	b.data = append(b.data, make([]byte, n)...)
+	return b.data[len(b.data)-n:]
+}
+
+// Data is unconsumed data remaining in the Buffer.
+func (b *Buffer) Data() []byte {
+	return b.data
+}
+
+// Remaining consumes and returns a copy of all remaining bytes in the Buffer.
+func (b *Buffer) Remaining() []byte {
+	p := b.Consume(len(b.Data()))
+	cp := make([]byte, len(p))
+	copy(cp, p)
+	return cp
+}
+
+// consume consumes n bytes from the Buffer. It returns nil, false if there
+// aren't enough bytes left.
+func (b *Buffer) consume(n int) ([]byte, bool) {
```

Comment:
this could be read, and the bool could be an error, and then return more info: 
return fmt.Errorf("Tried to read %d bytes but only %d available", ...